mayastor/tests/rebuild.rs

```diff
@@ -0,0 +1,1050 @@
+        get_nexus_state(nexus_hdl, NEXUS_UUID).await.unwrap(),
+        NexusState::NexusDegraded as i32
+    );
+
+    // Start rebuild.
+    start_rebuild(nexus_hdl, child).await.unwrap();
+    assert_eq!(get_num_rebuilds(nexus_hdl).await, 1);
+    assert!(wait_for_rebuild_state(
+        nexus_hdl,
+        child,
+        "running",
+        Duration::from_secs(1),
+    )
+    .await
+    .unwrap());
+}
+
+/// Test the rebuild progress gets updated.
+#[tokio::test]
+async fn rebuild_progress() {
+    let test = start_infrastructure("rebuild_progress").await;
+    let (mut ms1, _, ms3) = setup_test(&test, 1).await;
+    let nexus_hdl = &mut ms1;
+    let child = &get_share_uri(&ms3);
+
+    // Start a rebuild and give it some time to run.
+    add_child(nexus_hdl, child, true).await;
+    std::thread::sleep(Duration::from_millis(100));
+
+    // Pause rebuild and get current progress.
+    pause_rebuild(nexus_hdl, child).await;
+    assert!(wait_for_rebuild_state(
+        nexus_hdl,
+        child,
+        "paused",
+        Duration::from_secs(1),
+    )
+    .await
+    .unwrap());
+    let progress1 = get_rebuild_progress(nexus_hdl, child).await;
+
+    // Resume rebuild and give it some time to run.
+    resume_rebuild(nexus_hdl, child).await.unwrap();
+    std::thread::sleep(Duration::from_millis(100));
+
+    // Pause rebuild and check for further progress.
+    pause_rebuild(nexus_hdl, child).await;
+    let progress2 = get_rebuild_progress(nexus_hdl, child).await;
+    assert!(progress2 > progress1);
+}
+
+/// Test cases where a rebuild should not be started.
+#[tokio::test]
+async fn rebuild_not_required() {
+    let test = start_infrastructure("rebuild_not_required").await;
+    let (mut ms1, ms2, ms3) = setup_test(&test, 2).await;
+    let nexus_hdl = &mut ms1;
+    let child = &get_share_uri(&ms3);
+
+    // Attempt to rebuild a healthy child.
+    start_rebuild(nexus_hdl, child)
+        .await
+        .expect_err("Shouldn't rebuild");
+    assert_eq!(get_num_rebuilds(nexus_hdl).await, 0);
+
+    // Remove one of the healthy children.
+    remove_child(nexus_hdl, child).await;
+
+    // Can't rebuild a single child which is healthy.
+    let last_child = &get_share_uri(&ms2);
+    start_rebuild(nexus_hdl, last_child)
+        .await
+        .expect_err("Shouldn't rebuild");
+    assert_eq!(get_num_rebuilds(nexus_hdl).await, 0);
+}
+
+/// Test removing the source of a rebuild.
+#[tokio::test]
+async fn rebuild_src_removal() {
+    let test = start_infrastructure("rebuild_src_removal").await;
+    let (mut ms1, ms2, ms3) = setup_test(&test, 1).await;
+    let nexus_hdl = &mut ms1;
+    let child = &get_share_uri(&ms3);
+
+    // Pause rebuild for added child.
+    add_child(nexus_hdl, child, true).await;
+    pause_rebuild(nexus_hdl, child).await;
+    assert!(wait_for_rebuild_state(
+        nexus_hdl,
+        child,
+        "paused",
+        Duration::from_secs(1),
+    )
+    .await
+    .unwrap());
+    assert_eq!(
```

Comment:
You might want to put this check into a function with the expected state as a parameter

Reply:
Done :+1: